### DISCUSSION TITLE
When packaging 3rd-party xcframeworks as separate React Native Library, Embedded Framework is not being included

### DISCUSSION THREAD

**Discussion by silencer07** [callstack/react-native-builder-bob#439] [Q&A]
Apologies for long text. I need help so I am posting this and maybe someone better at me on iOS can help me.

Context:
We have 3rd party libraries(2 xcframeworks) in our app. They are proprietary framework that we cannot modify in anyway. This works currently.

I am tasked to pluck those frameworks out and package them as a react native library. I generated the library project using `npx create-react-native-library@latest` which is what is describe in the official react native documentation

I have already included and linked them in the generated project

![Screenshot 2023-08-16 at 9 01 59 AM](https://github.com/callstack/react-native-builder-bob/assets/19424606/ff8dc10b-a04d-4de9-9072-52b2935f3355)

![Screenshot 2023-08-16 at 9 02 09 AM](https://github.com/callstack/react-native-builder-bob/assets/19424606/3976fd96-278f-417b-95d0-7a421d97eab5)

![Screenshot 2023-08-16 at 9 03 43 AM](https://github.com/callstack/react-native-builder-bob/assets/19424606/f93f941c-bf2a-4afe-bbf1-176d3a2afe99)

Now I added the project as a dependency in the project and removed all the previously embedded xcframeworks (these two) since they should be coming from my new library that packages them. Note that we don't have CI/CD for the library so I am just including them as a raw git repo dependency in my package.json. The intent is to have it compiled together with the project when the app is building

However I am running on an issue, The `ServiceCore.xcframework` includes a framework folder inside it that contains some files, that for some reason I am having this issue

![Screenshot 2023-08-16 at 9 08 11 AM](https://github.com/callstack/react-native-builder-bob/assets/19424606/6bfaea43-d89d-4908-b624-e94ae80274a5)

I am seeing that SCCDefines inside the framework folder of the ServiceCore

![Screenshot 2023-08-15 at 5 05 01 PM](https://github.com/callstack/react-native-builder-bob/assets/19424606/257e242a-de86-495c-a12e-61a88f8cc15b)

I have two hunches:

Either the Chat framework does not see its sibling dependency during build, hence ServiceCore not found

Or Core xcframework not including its embedded framework during build.

I am pulling my hair on this one. If this is on android side I would not have any issues at all since gradle is good at dependency management.

[I also tried doing it altogether in podspec file and I am getting more error than the xcode way](https://github.com/callstack/react-native-builder-bob/discussions/292)

This is in the library framework btw

<img width="978" alt="Screenshot 2023-08-16 at 10 32 35 AM" src="https://github.com/callstack/react-native-builder-bob/assets/19424606/0c3101f2-f1f0-42f2-ae27-51d4be9dd95e">

I am having error

![Screenshot 2023-08-16 at 10 21 11 AM](https://github.com/callstack/react-native-builder-bob/assets/19424606/135ca164-c319-4cb2-a30a-2a256bc9ab2c)

The folder is wrong because it is looking for the `prepare-framework` file which is in the  _/Users/aldrin/WebstormProjects/mobile-app/**node-modules/react-native-mecca-chat-library**/ios/Frameworks/ServiceCore.xcframework_ (basically missing the node-modules/react-native-mecca-chat-library in path)

EDIT: Tried hardcoding the path, basically I am back to the error wherein the Core library is not found during build.

Thanks!

**Comment by atlj** [comment]
Hey, are you adding the `s.vendored_frameworks` line into the `.podspec` file or the `Podfile`?

**Reply by silencer07** [reply]
I tried adding it in podspec file, but i got the same error.

**Reply by atlj** [reply]
One idea I have is to also provide the path for that `sibling` framework on the `podspec` file.

```ruby
s.vendored_frameworks = 'ios/Frameworks/*.{xcframework}', 'ios/Frameworks/path/to/sibling/*.{framework}'
```

**Reply by silencer07** [reply]
I did tried that, but results to the same error because the header files are already “flattened”. I.e scc defines header file is the same level as the dependent sibling header file.

is there a way to instruct podspec file to retain the folder hierarchy when including xcframeworks?

**Reply by atlj** [reply]
Can you check this one? https://guides.cocoapods.org/syntax/podspec.html#preserve_paths

**Reply by silencer07** [reply]
I tried and I still have the same error. I already removed the references in xcode and put it in podspec.

this is my podspec file: https://pastebin.com/fQWNx4ba

This is the  node_module project inside the project that will use the library I am creating

![Screenshot 2023-08-25 at 7 51 03 AM](https://github.com/callstack/react-native-builder-bob/assets/19424606/c7047cf3-26ba-4bdb-9d3b-d74014417d99)

This is the resulting xcodeproj looks like when I view it on the dependent project in xcode

![Screenshot 2023-08-25 at 7 52 18 AM](https://github.com/callstack/react-native-builder-bob/assets/19424606/b9cc5fda-a471-4da7-83fa-6482788b6f5f)

**Reply by atlj** [reply]
The file [here]( https://pastebin.com/fQWNx4ba) is not a `podspec` file. It's a `Podfile`. Could you double check?

**Reply by silencer07** [reply]
@atlj I am pretty sure it is the spec file. This library does not generate a podfile, only podspec and xcodeproj file for iOS

Github is having issues atm I cannot upload a screenshot. I will link an imgur as proof that this is the generated podspec

**Reply by silencer07** [reply]
@atlj  here you go: https://imgur.com/a/z1V4EVu

**Reply by Noitham** [reply]
Hey @silencer07 , did you manage any progress? I have the exact same issue.

Exact same use case: 
Want to make a module for my react native app, the module depends on 2 SDK's provided via `.xcframework`.
Worked all good with the frameworks on the project, but want to extract it in a package to keep the project cleaner.

Seeing these errors when trying to build after installing the package:
![image](https://github.com/callstack/react-native-builder-bob/assets/32934478/4ee3d9de-5524-4852-bbd4-6bed19b7c8ff)

**Reply by silencer07** [reply]
@Noitham unfortunately no. We just decided to pause that task for now. Still I am interested if someone can make this work

**Reply by lampapos** [reply]
Try a workaround from [this issue](https://github.com/CocoaPods/CocoaPods/issues/6409) - separate the pod into two pods. One should contain only xcframeworks and another one should use it as a dependency.
